base/sort.jl

```diff
@@ -652,13 +669,203 @@ function sort!(v::AbstractVector, lo::Integer, hi::Integer, a::PartialQuickSort,
+#  * moves elements into the computed indices in the swap array
+#  * switches the swap and working array
+#
+# In the case of an odd number of passes, the returned list will === the input list t,
+# not v. This is one of the many reasons radix_sort! is not exported.
+function radix_sort!(v::AbstractVector{U}, lo::Integer, hi::Integer, bits::Unsigned,
+               ::Val{CHUNK_SIZE}, t::AbstractVector{U}) where {U <: Unsigned, CHUNK_SIZE}
+    # bits is unsigned and CHUNK_SIZE is a compile time constant for performance reasons.
+    MASK = UInt(1) << CHUNK_SIZE - 0x1
+    counts = Vector{UInt}(undef, MASK+2)
+
+    @inbounds for shift in 0:CHUNK_SIZE:bits-1
+
+        # counts[2:MASK+2] will store the number of elements that fall into each bucket.
+        # if CHUNK_SIZE = 8, counts[2] is bucket 0x00 and counts[257] is bucket 0xff.
+        counts .= 0
+        for k in lo:hi
+            x = v[k]                  # lookup the element
+            i = (x >> shift)&MASK + 2 # compute its bucket's index for this pass
+            counts[i] += 1            # increment that bucket's count
+        end
+
+        counts[1] = lo                # set target index for the first bucket
+        cumsum!(counts, counts)       # set target indices for subsequent buckets
+        # counts[1:MASK+1] now stores indices where the first member of each bucket
+        # belongs, not the number of elements in each bucket. We will put the first element
+        # of bucket 0x00 in t[counts[1]], the next element of bucket 0x00 in t[counts[1]+1],
+        # and the last element of bucket 0x00 in t[counts[2]-1].
+
+        for k in lo:hi # Is this iteration slower than it could be?
+            x = v[k]                  # lookup the element
+            i = (x >> shift)&MASK + 1 # compute its bucket's index for this pass
+            j = counts[i]             # lookup the target index
+            t[j] = x                  # put the element where it belongs
+            counts[i] = j + 1         # increment the target index for the next
+        end                           #  ↳ element in this bucket
+
+        v, t = t, v # swap the now sorted destination list t back into the primary list v
+
+    end
+
+    v
+end
+
+function sort!(v::AbstractVector{<:Bool}, lo::Integer, hi::Integer, a::AdaptiveSort, o::Ordering)
+    first = lt(o, false, true) ? false : lt(o, true, false) ? true : return v
+    count = 0
+    for i in lo:hi
+        if v[i] == first
+            count += 1
+        end
+    end
+    v[lo:lo+count-1] .= first
+    v[lo+count:hi] .= !first
+    v
+end
+
+maybe_unsigned(x::Integer) = x # this is necessary to avoid calling unsigned on BigInt
+maybe_unsigned(x::Union{Int8, Int16, Int32, Int64, Int128}) = unsigned(x)
+function _extrema(v::AbstractArray, lo::Integer, hi::Integer, o::Ordering)
+    mn = mx = v[lo]
+    while lo < hi
+        lo += 1
+        vi = v[lo]
```

Comment:
I guess `@inbounds` is OK here?
```suggestion
    @inbounds for i in lo:(hi-1)
        vi = v[i]
```

Reply:
`@inbounds` is okay, yes. I couldn't find any performance improvements from using it so I didn't. Looking deeper, at your prompting, I did find some. According to my benchmarks the following three versions have a runtime of 1ns per element for `rand(10_000)` under the ordering `Base.Sort.Float.Left()`:
```julia
while lo < hi
        lo += 1
        vi = v[lo]
```
```
@inbounds while lo < hi
        lo += 1
        vi = v[lo]
```
```
@inbounds for i in lo+1:hi
        vi = v[i]
```
and this one has a runtime of 1.2ns per element:
```
for i in lo+1:hi
        vi = v[i]
```
When computing the `_extrema` of 10_000 Int64s, those with the `@inbounds` annotation ran at 0.20ns, and those without ran at 0.23ns. Based on this, I accept your recommendation (with indexing fix) as an improvement in readability and speed. (aside: 0.20ns is the fastest benchmarking result I have ever seen. The `@code_llvm` instructions are pretty complex.)
```suggestion
    @inbounds for i in (lo+1):hi
        vi = v[i]
```

Reply:
less than 1ns means the compiler broke your benchmark

Reply:
My methodology is
```julia
(@belapsed Base.Sort._extrema(x,1,10_000,Base.Forward) setup=(x=rand(Int,10_000)))/10_000
# 2.0268e-10
```
My conclusion is that the runtime _per element_ is `.20ns`. I believe the compiler optimized effectively but did not break the benchmark.

Reply:
Strangely, my implementation seems to be slightly faster than `extrema`. Perhaps this has something to do with the shortcuts I took assuming that the input was non-empty?
Methodology:
```julia
@belapsed extrema(x) setup=(x=rand(Int,10_000))
# 2.4671111111111115e-6
@belapsed Base.Sort._extrema(x, 1, 10_000, Base.Forward) setup=(x=rand(Int,10_000))
# 2.0268e-6
```

Reply:
I moved this conversation to https://github.com/JuliaLang/julia/issues/44606